docs/adrs/004-deprecating-and-removing-flags.md

```diff
@@ -0,0 +1,34 @@
+# 4. Deprecating and Removing Flags
+
+Date: 2022-07-20
+
+## Status
+
+Accepted
+
+## Context
+
+Currently we have several flags that are no longer [used](https://k3s-io.github.io/docs/reference/server-config#deprecated-options). These have been labelled as deprecated, but we currently lack a system around removing them completely. This creates a problem where we aren't really deprecating flags, because they are always kept around.
+
+## Decision
+
+The following system will be implemented for depcrecating and removing flags:
+
+1) Flags can be declared as deprecated at any time.
+2) Flags that are deprecated must be labeled as such on all currently supported releases. 
+3) Once a flag has been marked as deprecated, it will be converted to a hidden flag on the next patch release. E.g. `--foo` will be marked as deprecated in v1.24.2, and will be converted to a hidden flag in v1.24.3.
+   Additionally, the flag will begin to warn users that it is going to be removed in the next minor release. E.g. `--foo` will cause a warning in v1.24.3, telling users that "--foo will be removed in v1.25.0".
+4) In the next minor release branch, flags deprecated will become "nonoperational" in that they will cause a fatal error if used. This error must exlapin to the user any new flags or configuration that replace this flag. 
```

Comment:
The historical policy from Craig was that we needed to maintain flag compatibility for two minor releases before making any breaking changes. So:
1. Supported in 1.23
2. Marked deprecated in 1.24
3. Warning but continues to work in 1.25
4. Removed in 1.26

Reply:
Reading your edit, I have modified the policy. Now we can decide that a flag "will be deprecated" on any minor patch, and will begin warning people about it. But we will not move that flag into "deprecated" until the next major release. I want to be able to warn users ASAP so they have the most amount of time to change their config.

Reply:
Here's the upstream flag deprecation policy, which I suggest we align ourselves with: https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli